Sources/ApplicativeRouter/Combinators.swift

```diff
@@ -0,0 +1,190 @@
+public func jsonBody<A: Codable>(_ type: A.Type) -> Router<A> {
+  return dataBody.map(PartialIso.codableToData.inverted)
+}
+
+/// Parses the end of the request data by making sure that all of the path components have been consumed.
+public let _end = Router<Prelude.Unit>(
+  parse: { route in
+    route.path.isEmpty
+      ? (RequestData(method: route.method, path: [], query: [:], body: nil), unit)
+      : nil
+},
+  print: const(.empty),
+  template: const(.empty)
+)
+
+extension Router {
+  /// Parses and consumes a path component as an int.
+  public static var int: Router<Int> { return pathParam(.int) }
+
+  /// Parses and consumes a path component as a bool.
+  public static var bool: Router<Bool> { return pathParam(.bool) }
+
+  /// Parses and consumes a path component as a string.
+  public static var string: Router<String> { return pathParam(.id) }
+
+  /// Parses and consumes a path component as a double.
+  public static var num: Router<Double> { return pathParam(.double) }
+}
+
+/// Parses the query params to create a value of type `A` via the `Codable` protocol.
+/// TODO: this only works for types `A` with string fields. Is it possible to improve that?
+public func queryParams<A: Codable>(_ type: A.Type) -> Router<A> {
+  return .init(
+    parse: { route in
+      (try? JSONSerialization.data(withJSONObject: route.query))
+        .flatMap { try? JSONDecoder().decode(A.self, from: $0) }
+        .map { (route, $0) }
+  },
+    print: { a in
+      let params = (try? JSONEncoder().encode(a))
+        .flatMap { try? JSONSerialization.jsonObject(with: $0) }
+        .flatMap { $0 as? [String: Any] }
+        .map { $0.mapValues { "\($0)" } }
+        ?? [:]
+      return RequestData(method: nil, path: [], query: params, body: nil)
+  },
+    template: { a in
+      let params = (try? JSONEncoder().encode(a))
+        .flatMap { try? JSONSerialization.jsonObject(with: $0) }
+        .flatMap { $0 as? [String: Any] }
+        .map { $0.mapValues { _ in ":string" } }
+        ?? [:]
+      return RequestData(method: nil, path: [], query: params, body: nil)
+  })
+}
+
+/// Parses the HTTP method verb of the request.
+public func method(_ method: Method) -> Router<Prelude.Unit> {
+  return Router(
+    parse: { route in
+      route.method == method
+        ? (route |> \.method .~ nil, unit)
+        : nil
+  },
+    print: { _ in  .init(method: method, path: [], query: [:], body: nil) },
+    template: { _ in  .init(method: method, path: [], query: [:], body: nil) }
+  )
+}
+
+public let delete = method(.delete)
+public let get = method(.get)
+public let head = method(.head)
+public let options = method(.options)
+public let patch = method(.patch)
+public let post = method(.post)
+public let put = method(.put)
+
+// MARK: - Experimental
+
+// TODO: open for discussion
+infix operator <=>: infixr9
+
+public func <=> <A> (lhs: String, rhs: PartialIso<String, A>) -> Router<A> {
```

Comment:
im not sure how deep into custom operators we wanna go for routing. for example this allows us to do `"ref" <=> opt(.string)` instead of `pathParam("ref", opt(.string))`, etc. we could also have things like `</>`, `<?>`, `<&>`, and if we dod the state machine phantom types we'd be able to make sure the operators only work where we expect.